[pytorch][PR] Record FutureNCCL callback stream on CUDA caching allocator

wayi1 opened the pull request:
Summary:
When calling `then()` from WorkNCCL, record the input data pointers in futureNCCLCallbackStream_ before the execution of the input callback.

Note that the recording cannot be directly added to the lambda used by addCallback in ProcessGroupNCCL.hpp. This is because the type of future value in that context is pyobject rather than TensorList, but a type casting will require pybind and introduce Python dependency, which should not be allowed in c10d library.

I have considered creating a util function in a separate file to support this type casting, and then placing it under torch/csrc directory where python dependency is allowed. However, torch/csrc has a dependency on c10d, so this will create a circular dependency.

Finally, a `record_stream_cb_` member is added to FutureNCCL, and the default value is nullptr. A default `record_stream_cb_` implementation is added to `caffe2/torch/csrc/jit/python/pybind_utils.h`, where Python dependency is allowed. 

Currently record_stream_cb_ is not really customizable, and we assume the input ivalue of this callback is or can be casted into a tensor that has exactly one tensor. This callback will only be used by NCCL backend.

Test Plan
buck test @mode/dev-nosan caffe2/test/distributed:c10d -- ProcessGroupNCCLTest
buck test @mode/dev-nosan caffe2/test/distributed:c10d  -- test_accumulate_gradients_no_sync_allreduce_with_then_hook
buck test @mode/dev-nosan caffe2/test/distributed:c10d  -- test_ddp_comm_hook_allreduce_with_then_hook_nccl

Differential Revision: D23910257

pritamdamania87 commented:
@ngimel Could you help review this PR? We're following the convention from here: https://github.com/pytorch/pytorch/blob/master/torch/lib/c10d/ProcessGroupNCCL.cpp#L968

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.cpp:
```diff
@@ -916,6 +918,13 @@ c10::intrusive_ptr<c10::ivalue::Future> ProcessGroupNCCL::WorkNCCL::
       outputs_->size() == 1,
       "WorkNCCL's getFuture API is only supported for single-process single-device mode.");
   auto deviceIndex = (*outputs_)[0].device().index();
+  // Do not free the resluting tensors before their usages on
+  // futureNCCLCallbackStream_ finish.
+  for (const auto& tensor : *outputs_) {
+    c10::cuda::CUDACachingAllocator::recordStream(
+        tensor.storage().data_ptr(), *futureNCCLCallbackStreams_[deviceIndex]);
+  }
```
Don't think we need recordStream here since we aren't actually running on the futureNCCLCallbackStream here.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -295,6 +296,13 @@ class ProcessGroupNCCL : public ProcessGroup {
           [&](std::function<at::IValue(void)> cb) {
             try {
+              // Do not free the input tensors before their usages on
+              // futureNCCLCallbackStream_ finish.
```
There is only one tensor in the vector, so let's just do something like this to be explicit:

```
auto tensors = value_.toTensorVector();
TORCH_INTERNAL_ASSERT(1, tensors.size());
c10::cuda::CUDACachingAllocator::recordStream(
                    tensors[0].storage().data_ptr(), *futureNCCLCallbackStream_);
```

SciPioneer replied:
Done. The size of tensors is indeed 1.

However, realized that the change here cannot pass test_accumulate_gradients_no_sync_allreduce_with_then_hook which invokes then -- the execution fails after this , with an error of "Timing out after 100 seconds and killing subprocesses".

I even tried to place this added code after "fut->markCompleted(at::IValue(cb()));" or even right before returning the future in this lambda, and tried to comment out recordStream call. Didn't work.

pritamdamania87 replied:
Can you debug to see why the test is timing out? You can probably add some log statements to the unit test to see where it is getting stuck. Another option is to attach gdb to the child processes when the test is stuck.

SciPioneer replied:
I added LOG statements to debug before asking actually, but all the logs in this function could be printed out, so it's not timed out in this function. Will try gdb.

SciPioneer replied:
It turns out that the type of future value here is pyobject rather than TensorList, but a type casting will require pybind and introduce Python dependency, which should not be allowed in c10d library.

I have considered creating a util function in a separate file to support this type casting, and then placing it under torch/csrc directory where python dependency is allowed. However, torch/csrc has a dependency on c10d, so this will create a circular dependency.

Per offline discussion, added a `record_stream_cb_` member to FutureNCCL, and the default value is nullptr. A default `record_stream_cb_` implementation is added to `get_future` binding in init.cpp, where Python dependency is allowed. record_stream_cb_ can be customized by caller via `setRecordStreamCallback`, so here the future value does not have to be a tensor vector.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -520,7 +528,10 @@ Example::
       .def(py::init<>());
 
-  shared_ptr_class_<::c10d::TCPStore>(module, "TCPStore", store,
+  shared_ptr_class_<::c10d::TCPStore>(
+      module,
+      "TCPStore",
+      store,
```
nit: Usually try to avoid formatting changes since it makes the PR harder to review. You can always send a separate PR with the formatting changes.

SciPioneer replied:
Yeah, realized that. Will be more careful next time.

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
+                  if (value.isTensorList()) {
+                    tensors = std::move(value.toTensorVector());
+                  } else {
+                    py::object obj = torch::jit::toPyObject(value);
```
We probably need to acquire GIL only for this codepath? Can we move `pybind11::gil_scoped_acquire gil;` here?

pritamdamania87 commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
+              };
+              fut->setRecordStreamCallback(record_stream_cb);
+            }
+            fut->recordStream();
```
recordStream should not be called here, it should be done only just before we call execute an OP on a new stream. Basically it should be done just before `streamGuard` in `addCallback()`

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
       return !value_.isNone();
     }
 
+    std::shared_ptr<at::cuda::CUDAStream> getfutureNCCLCallbackStream() {
```
This method is not used.

pritamdamania87 commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
+      record_stream_cb_ = std::move(record_stream_cb);
+    }
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
+      // Do not free the underlying data storage of value_ before its usage on
+      // futureNCCLCallbackStream_ finish.
+      record_stream_cb_(value_, *futureNCCLCallbackStream_);
+    }
```
You probably don't need a separate method for this if you're doing it in `addCallback()` instead.

pritamdamania87 reviewed (changes requested):
Requesting changes since I think `recordStream` should be done inside addCallback(). Also, can we run the DDP/HPC benchmarks with this change (with fp16 hook enabled) to ensure it doesn't cause any issues? You can probably verify QPS and peak memory to see they're not affected significantly.

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
+              auto record_stream_cb = [](const at::IValue& value,
+                                         at::cuda::CUDAStream cuda_stream) {
+                if (value.isTensorList() || value.isPyObject()) {
+                  pybind11::gil_scoped_acquire gil;
```
I think this function already holds GIL (because there is no gil_scoped_release guard in the definition)? You can probably check this by using `PyGILState_Check`. If so, I think re-acquiring the GIL here could result in some issues.

SciPioneer replied:
The new design moved this part to pybind_utils.h

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
+                            obj, c10::ListType::create(c10::TensorType::get()))
+                            .toTensorVector());
+                  }
+                  TORCH_INTERNAL_ASSERT(1, tensors.size());
```
Is this the correct use of the API? Shouldn't it be `TORCH_INTERNAL_ASSERT(tensors.size() == 1, "expected tensors.size to be 1")` or something like that?

SciPioneer replied:
Fixed.

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
           "get_future",
           [](::c10d::ProcessGroup::Work& work)
               -> std::shared_ptr<jit::PythonFutureWrapper> {
+#ifdef USE_C10D_NCCL
```
Might just be my confusion on this compiler guard. But isn't this true if we compile with NCCL enabled, but the user could still choose the Gloo backend at runtime? What sort of error message do we get if this codepath is executed with the Gloo backend? Could we add a test to ensure something reasonable is raised in this case (can be in a separate diff).

SciPioneer replied:
There shouldn't be any side effect on other backend, as no code will be executed.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
+    }
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
```
nit: I believe this should be `TORCH_INTERNAL_ASSERT` as it is a framework bug if this function is called with an invalid record_stream_cb_. 

I believe TORCH_CHECK is used for error validation coming from user input/code, and TORCH_INTERNAL_ASSERT for framework validation.

SciPioneer replied:
Fixed.

rohan-varma commented on torch/csrc/distributed/c10d/init.cpp:
```diff
@@ -992,6 +1010,36 @@ that adds a prefix to each key inserted to the store.
+            // By default record_stream_cb_ of the future is not set, and here
+            // we assume the type of this future value is or can be casted into
+            // a tensor vector that only has exactly one tensor.
+            if (!fut->hasRecordStreamCallback()) {
```
Could there be race conditions where different threads could try to set this at the same time? Do we support user calling this code from multithreaded environment?

SciPioneer replied:
The asynchronous threads/streams on the same device actually will be chained in the order of creation time (i.e., forked by the call of then). See the comment on addCallback function in ProcessGroupNCCL.hpp.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
     c10::DeviceIndex deviceIndex_;
     std::shared_ptr<std::vector<at::cuda::CUDAEvent>> cudaEvents_;
     std::shared_ptr<at::cuda::CUDAStream> futureNCCLCallbackStream_;
+    std::function<void(const at::IValue&, at::cuda::CUDAStream)>
```
Nit: up to your preference, but you could do `using recordStreamCallback = std::function<void (....` to avoid having to write this twice. Although since it's used only twice not sure if it's worth it.

SciPioneer replied:
I also had the same thought. After realizing that `std::function<at::IValue(void)> cb` has also appeared twice in this file, I decided not to create a separate alias.

rohan-varma commented on torch/lib/c10d/ProcessGroupNCCL.hpp:
```diff
@@ -333,11 +333,34 @@ class ProcessGroupNCCL : public ProcessGroup {
+
+    void recordStream() {
+      TORCH_CHECK(record_stream_cb_);
+      // Do not free the underlying data storage of value_ before its usage on
```
cc @pritamdamania87 , is there anyway we could write a test that would fail without this record stream callback logic?

pritamdamania87 replied:
I think it might be possible but tricky to do. It probably has to be a C++ test and interact with CUDACachingAllocator in a way that the caching allocator frees up the data, but we're still using the data in futureNCCLCallbackStream.

SciPioneer replied:
Currently I hacked the test file to locally verify that in a test case like fut.then(mult).then(div), CUDACachingAllocator::recordStream has been invoked by the first two futures, respectively.

Agree that there could be a better approach to create a C++ unit test that can allow CUDACachingAllocator to free up the cached data.

SciPioneer commented:
> Requesting changes since I think `recordStream` should be done inside addCallback(). Also, can we run the DDP/HPC benchmarks with this change (with fp16 hook enabled) to ensure it doesn't cause any issues? You can probably verify QPS and peak memory to see they're not affected significantly.

Per offline discussion, ran the callback right before `addCallback` is invoked.

pritamdamania87 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -411,6 +411,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
 
+  // Since this file cannot import CUDA depedency, use void* to represent CUDA
+  // stream, and will cast it into the correct stream type whenever use it.
+  virtual void setRecordStreamCallback(
```
You can change the signature of this method to the following to address the CUDA stream issue:

```
template< typename T>
virtual void setRecordStreamCallback(T record_stream_cb) {}
```

This is also slightly more efficient than std::function and I think you can pass in a lambda here for `record_stream_cb` which accepts a CUDA stream as input.

SciPioneer replied:
Please note that setRecordStreamCallback is defined inside ivalue::Future. Do we want to make ivalue::Future a template class? If so, how shall we instantiate T for other backends like Gloo and MPI?

Another concern: I am a fan of template class, but making this core class as a template class for only such a minor method may not justify the loss of debug-ability to some extent, due to the notorious error messages. As I can see, this class seems to be updated quite a few times recently.

SciPioneer replied:
Alternatively, what about replacing void* and reinterpret_cast with std::any and any_cast? This is a C++17 feature, and I am glad that our codebase can support C++17.

SciPioneer replied:
Per offline discussion, used c10::Stream to construct CUDAStream on the fly and avoid the usage of reinterpret_cast.

pritamdamania87 commented on aten/src/ATen/core/ivalue_inl.h:
```diff
@@ -411,6 +411,11 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     return fut;
   }
 
+  // Since this file cannot import CUDA depedency, use void* to represent CUDA
+  // stream, and will cast it into the correct stream type whenever use it.
```
Add a comment here mentioning the default implementation is a no-op since it doesn't deal with any CUDA streams.

pritamdamania87 reviewed (commented):
Changes LGTM, although there are a lot of CI failures. Can you fix those?

Commit:
[pytorch][PR] Record FutureNCCL callback stream on CUDA caching allocator (#45318)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/45318

When calling `then()` from WorkNCCL, record the input data pointers in futureNCCLCallbackStream_ before the execution of the input callback.

Note that the recording cannot be directly added to the lambda used by addCallback in ProcessGroupNCCL.hpp. This is because the type of future value in that context is pyobject rather than TensorList, but a type casting will require pybind and introduce Python dependency, which should not be allowed in c10d library.

I have considered creating a util function in a separate file to support this type casting, and then placing it under torch/csrc directory where python dependency is allowed. However, torch/csrc has a dependency on c10d, so this will create a circular dependency.

Finally, a `record_stream_cb_` member is added to FutureNCCL, and the default value is nullptr. A default `record_stream_cb_` implementation is added to `PythonFutureWrapper,` where Python dependency is allowed.

In addition, a few lines are reformatted by lint.
caffe2/torch/csrc/distributed/c10d/init.cpp is only reformatted.

#Closes: https://github.com/pytorch/pytorch/issues/44203

Test Plan:
buck test mode/dev-nosan caffe2/test/distributed:c10d -- ProcessGroupNCCLTest
buck test mode/dev-nosan caffe2/test/distributed:c10d  -- test_accumulate_gradients_no_sync_allreduce_with_then_hook
buck test mode/dev-nosan caffe2/test/distributed:c10d  -- test_ddp_comm_hook_allreduce_with_then_hook_nccl

Reviewed By: pritamdamania87

Differential Revision: D23910257

fbshipit-source-id: fe41338e3b5510b9d9a631f57b78134e8c7e1b12
